### PR TITLE
Offline Mode: Autosave and State Restoration

### PR DESCRIPTION
**Pull request by kean**
Update autosave according to the spec.

## Testing

**Test 1.1**

- Create and publish a new post on wp.com
- Make some changes and wait until it creates an [autosave](https://jetpack.com/blog/wordpress-autosave/)
- Open the app
- Open the post created on step 1
- **Verify** that it shows an "Autosave Available" alert with teh date when the autosave was created and with "View Changes" and "Cancel" button.
- Perform tests 1.1.x

**Test 1.1.1 (Cancel)**

- Tap "Cancel"
- **Verify** that nothing changes in the editor
- Tap "Back"
- **Verify** that the editor was dismissed without a confirmation dialog
- Re-open the post
- **Verify** that is shown an "Autosave Available" alert again

**Test 1.1.2 (View and Discard)**

- Tap "View Changes"
- **Verify** that the autosaved content was loaded in the editor
- Tap "Back"
- **Verify** that the confirmation dialog was shown
- Tap "Discard Changes"
- Re-open the post
- **Verify** that is shown an "Autosave Available" alert again

**Test 1.1.2 (View and Update)**

- Tap "View Changes"
- **Verify** that the autosaved content was loaded in the editor
- Tap "Update"
- **Verify** that the post was updated
- Tap "Back"
- **Verify** that no confirmation dialog was shown
- Re-open the post
- **Verify** that no "Autosave Available" dialog was shown

### State Restoration

**Test 2.1**

- Create a new post
- Terminate the app while the editor is opened
- Re-open the app
- **Verify** that the editor was opened automatically (state restoration)
- **Verify** that you can publish a post

**Test 2.2**

- Open an existing post and make changes
- Terminate the app while the editor is opened
- Re-open the app
- **Verify** that the editor was opened automatically (state restoration)
- **Verify** that you can publish a post

## Regression Notes
1. Potential unintended areas of impact

2. What I did to test those areas of impact (or what existing automated tests I relied on)

3. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding unit tests for my changes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WordPress/Classes/ViewRelated/Post/EditPostViewController.swift [hunk 1]
```diff
@@ -244,7 +245,7 @@ extension EditPostViewController: UIViewControllerRestoration {
                 return nil
         }
 
-        return EditPostViewController(post: reloadedPost)
+        return EditPostViewController(post: reloadedPost.latest() as! Post)
```
Comment (kean) [inline WordPress/Classes/ViewRelated/Post/EditPostViewController.swift:248]: A workaround for an issue caught by the following assertion:

```
assert(post.latest() == post, "Must be opened with the latest verison of the post")
```

File: WordPress/Classes/ViewRelated/Post/PostEditor+Publish.swift [hunk 2]
```diff
@@ -514,9 +514,6 @@ extension PublishingEditor {
         alert.addDestructiveActionWithTitle(Strings.closeConfirmationAlertDiscardChanges) { _ in
             self.discardAndDismiss()
         }
-        alert.addActionWithTitle(Strings.closeConfirmationAlertSaveDraftChanges, style: .default) { _ in
```
Comment (kean) [inline WordPress/Classes/ViewRelated/Post/PostEditor+Publish.swift:517]: You'll have to save the changes – same in production. We'll collect the feedback and see if there needs to be some form of autosave like on the web.

**Review by momo-ozawa — COMMENTED** [review]
I couldn't get state restoration to work - when I terminate the app and reopen it, the editor isn't displayed

**Comment by kean** [comment]
> I couldn't get state restoration to work - when I terminate the app and reopen it, the editor isn't displayed

I tested it with the WP app with the latest migration phase where there is no MySite screen. It looks like it fails to restore:

```swift
    func application(_ application: UIApplication, viewControllerWithRestorationIdentifierPath identifierComponents: [String], coder: NSCoder) -> UIViewController? {
        guard let restoreID = identifierComponents.last else {
            return nil
        }

        return Restorer().viewController(identifier: restoreID)
    }
```

ID to look for is "MySiteNavigationControllerRestorationID".

It looks like none of the restoration code has been working in the app (and nobody noticed).

**Review by momo-ozawa — APPROVED** [review]

**Pull request merged**